src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -423,50 +424,97 @@ private TypeSymbol NarrowVariableAccessType(TypeValidatorConfig config, Variable
             return targetType;
         }
 
+
         private TypeSymbol NarrowUnionType(TypeValidatorConfig config, SyntaxBase expression, TypeSymbol expressionType, UnionType targetType)
+            => expressionType switch
+            {
+                UnionType expressionUnion => NarrowUnionTypeForUnionExpressionType(config, expression, expressionUnion, targetType),
+                _ => NarrowUnionTypeForSingleExpressionType(config, expression, expressionType, targetType),
+            };
+
+        private TypeSymbol NarrowUnionTypeForUnionExpressionType(TypeValidatorConfig config, SyntaxBase expression, UnionType expressionType, UnionType targetType)
         {
             List<UnionTypeMemberViabilityInfo> candidacyEvaluations = new();
 
-            foreach (var candidateType in targetType.Members)
+            foreach (var candidateExpressionType in expressionType.Members)
             {
-                if (!AreTypesAssignable(expressionType, candidateType.Type))
+                if (!AreTypesAssignable(candidateExpressionType.Type, targetType))
                 {
-                    candidacyEvaluations.Add(new(candidateType, null, ImmutableArray<IDiagnostic>.Empty));
+                    candidacyEvaluations.Add(new(candidateExpressionType, null, ImmutableArray<IDiagnostic>.Empty));
                     continue;
                 }
 
                 var candidateDiagnostics = ToListDiagnosticWriter.Create();
                 var candidateCollector = new TypeValidator(typeManager, binder, candidateDiagnostics);
-                var narrowed = candidateCollector.NarrowType(config, expression, candidateType.Type);
+                var narrowed = candidateCollector.NarrowUnionTypeForSingleExpressionType(config, expression, candidateExpressionType.Type, targetType);
```

Comment:
Is there any harm in using `NarrowUnionType` here (just in case we have a `UnionType` with a `UnionType` member)?